pythonwhat/probe.py

```diff
@@ -0,0 +1,218 @@
+
+    def __call__(self, state=None):
+        """Call original function with its arguments, and optional state"""
+        ba = self.data['bound_args']
+        self.data['func'](state=state, *ba.args, **ba.kwargs)
+
+    def partial(self):
+        """Return partial of original function call"""
+        return partial(self.data['func'], **self.data['bound_args'].arguments)
+
+    def update_child_calls(self):
+        """Replace child nodes on original function call with their partials"""
+
+        for node in filter(lambda n: len(n.arg_name), self.child_list):
+            self.data['bound_args'].arguments[node.arg_name] = node.partial()
+
+    def remove_child(self, node):
+        indx = self.child_list.index(node)
+        del self.child_list[indx]
+        return indx
+
+    def add_child(self, child):
+        # since it is a tree, there is only one parent 
+        # note this means we do not allow edges between same layer units
+        if child.parent: child.parent.remove_child(child)
+        child.parent = self
+        self.child_list.append(child)
+
+    @property
+    def depth(self):
+        if self.parent: return self.parent.depth + 1
+        else: return 0
+
+    def __str__(self):
+        # TODO print function signature without defaults (or with)
+        return pp.pformat(self.data)
+
+    def __iter__(self):
+        for c in self.child_list: yield c
+
+class NodeList(Node):
+    def partial(self):
+        return [node.partial() for node in self.child_list]
+
+    def update_child_calls(self):
+        pass
+
+class NodeDict(Node):
+    def partial(self):
+        return OrderedDict((node.arg_name, node.partial()) for node in self.child_list)
+
+    def update_child_calls(self):
+        pass
+
+class Probe(object):
+    def __init__(self, tree, f):
+        self.tree = tree
+        self.f = f
+        self.test_name = f.__name__
+        # TODO: auto sub_test detection
+        self.sub_tests = SUB_TESTS.get(self.test_name) or []
+    
+    def __call__(self, *args, **kwargs):
+        """Bind arguments to original function signature, and store in a Node
+
+        This is used to discover what tests and sub-tests the SCT would like to 
+        call, and defer them for later execution via their node instance. Node
+        instances are assembled into a tree.
+
+        """
+
+        bound_args = inspect.signature(self.f).bind(*args, **kwargs)
+
+        data = dict(
+                bound_args = bound_args, 
+                func = self.f)
+        this_node = Node(data=data, name=self.test_name)
+        self.tree.crnt_node.add_child(this_node)
+
+        da = bound_args.arguments
+        for st in self.sub_tests:     # TODO: auto sub test detection
+            if st in da and da[st]:
+                self.run_sub_tests(da[st], self.tree, this_node, st)
+        return this_node
+
+    @staticmethod
+    def run_sub_tests(test, tree, node, arg_name):
```

Comment:
Is `run_sub_tests()` a good name? It's not actually running tests, is it? It's _building_ nodes?